### ISSUE TITLE
Singing synthesis opcodes

### ISSUE THREAD

**Issue by gogins** [csound/csound#658]
Implement an opcode or opcodes for vocal singing synthesis in Csound, inspired by Vocaloid, Sinsy, and such.

The opcode should take marked up text and some form of vocal model, and synthesize a musically usable facsimile of human vocals with correct time and pitch. It need not be perfect but it must be musically useful.

The general issues in this project are:
1. Singing synthesis is not a trivial task.
2. Singing synthesis has commercial value, so most implementations that actually work are closed source.
3. Singing synthesis is a fad in Japan but not very popular elsewhere, so most open source development takes place in Japan for the Japanese language only.

**Comment by gogins** [comment]
Then there is Festival [http://www.cstr.ed.ac.uk/projects/festival/].

**Comment by gogins** [comment]
Check this out: https://github.com/Sleepwalking/libllsm.

**Comment by gogins** [comment]
It is becoming a bit clearer what is going on in this subculture. Mostly the people are otaku. Their interest is in making song videos with sexy anime girl singers. The software licensing issues are bad because there is some commercial potential. So when something starts to work well it quits being available. In addition development of the Flinger singer for Festival stopped when Mike Macon died. So far I haven't been able to find source code for Flinger. I am still looking.

In a nutshell Vocaloid works fairly well but is proprietary. Utauloid is a free software clone of Vocaloid but works only for Japanese voices. Also utaloid is not a standard package and is not that easy to get. Isengaara's package qtau tries to bring all this stuff together for English voices. To build it clone qtau and then run qtau/get_utau/TETO and qtau/depends/build_depends. Other dependencies revealed by trying to build qtau can be obtained from public package and code repositories and installed, or built and installed.

When I get some software that sings that I can build, the fun begins. I have to rewrite it, package it, cut out stuff not needed, and develop the interface as opcodes.

**Comment by gogins** [comment]
Road block:
dget: retrieving http://mentors.debian.net/debian/pool/main/v/vconnectstand/vconnectstand_20151108a-1.dsc
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   381    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found

Then https://github.com/cadencii/vConnect.git has a Windows only build system.

**Comment by gogins** [comment]
The Universitat Pompeu Fabra Music Technology Group (under Xavier Serra) appears to be a big part of Sony's Vocaloid. No freebies here. This appears to be applied research for hire.

**Comment by gogins** [comment]
# Status

1. It is clear that several teams have developed singing voice synthesis to the point of being quite useful for a computer music system such as Csound including at least the Festival singer, Utau, the Moresampler, and Sinsy.

2. Unfortunately the Festival singer is not open source, Utau is not open source, Kanru Hua's Moresampler does not do English and is not open source although its libraries are, and Sinsy is open source but does not do English.

Until such time as at least one of the projects in (2) becomes usable (open source AND English), I am placing this project on hold.

**Comment by gogins** [comment]
Isengaara responded to my issues. He removed references to utau files and clarified that espeak rebuilt with singing mode is the only plugin required. I am going to try again.

**Comment by gogins** [comment]
Unfortunately Isengaara is using a freeware but closed source synthesizer now (madde).

**Comment by gogins** [comment]
I am checking out http://www.isca-speech.org/archive/Interspeech_2016/.

**Comment by gogins** [comment]
More news: a variant of qtau here https://github.com/qtau-devgroup.

**Comment by gogins** [comment]
I offered to help Kanru Hua with MoreSampler so I could use it in a Csound plugin. He finally responded with this:

Hello Michael,

Thanks for your interest in Moresampler.
I feel sorry that I might not have enough time to work on a side project, but hopefully the open source parts or wheels I built for Moresampler could be useful. The links on my blog are a bit outdated. Here's a newer summary,
speech signal processing snippets, https://github.com/Sleepwalking/ciglet
fundamental frequency tracking, https://github.com/Sleepwalking/libpyin
variants of Viterbi algorithm, https://github.com/Sleepwalking/libgvps
left-to-right hidden semi-Markov models (good tool for automatic text-to-speech alignment), https://github.com/Sleepwalking/liblrhsmm
The libllsm repository is basically deprecated. Quite some flaws were found in the algorithm and I'm still in the half way of redesigning. For speech analysis/synthesis I'd recommend you to have a look at Dr. Morise's World library (https://github.com/mmorise/World), which recently has been made real-time.

Aside from signal processing, the following papers on the data-driven/machine-learning approaches to speech synth would also be indispensable to a full-fledged synthesizer:
K Saino, H Zen, Y Nankaku, A Lee, K Tokuda. An HMM-based singing voice synthesis system. 
AJ Hunt, AW Black. Unit selection in a concatenative speech synthesis system using a large speech database.
K. Tokuda, et al. Speech synthesis based on hidden Markov models.
While the current trend in the field is, of course, deep learning based synthesis, at this point the hardware/software infrastructure doesn't seem to be accessible enough. Google and Microsoft keep publishing papers just for conquering academic+commercial territory. So you'd probably rather want to base your system on a slightly old-school fashioned library, e.g. flite (http://www.festvox.org/flite/), HTS (http://hts.sp.nitech.ac.jp/). I understand that HTK license sucks but HTS does have a synthesis-only BSD-licensed version called hts_engine. Though I haven't done so, it is possible to train a synthesizer using liblrhsmm.

Good luck on your Csound plugin!

Regards,
Kanru Hua

I don't think he has any interest in a completely open source program, but I asked again more directly.

**Comment by gogins** [comment]
I'm going to dink around with World and with qtau-devgroup.

**Comment by gogins** [comment]
An alternative approach to singing voice synthesis has been developed based on working with spectra directly instead of filters: http://asa.scitation.org/doi/abs/10.1121/1.4969144 and https://arxiv.org/pdf/1704.03809.pdf.

**Comment by gogins** [comment]
Lots of proprietary research at FPU and other places, based on DNN and RNN, plus GPU processing.

In the meantime I am evaluating existing commercial plugins. That might work if Csound can send MIDI to them via vst4cs.

[Cantor](http://www.virsyn.de/en/E_Products/E_CANTOR/e_cantor.html)

**Comment by gogins** [comment]
[Here](https://news.ycombinator.com/item?id=7875612) is a helpful discussion of Utau which has English voice banks now.

**Comment by gogins** [comment]
I keep googling but the source code doors are not opening, and the commercial software either is foreign or is not good enough. I am closing this issue in defeat, but may re-open it.

**Comment by jamiebullock** [comment]
Thanks for posting this. I've just been down a similar rabbit hole myself, and reached similar conclusions.

Since you posted there is now [Synthesiser V](https://synthesizerv.com/en/) which seems pretty good but a completely closed project.

There is also some [interesting work](https://github.com/numediart/mage) at Numediart, but this is GPL (although the underlying libraries seem to be BSD)

I suspect the future of this lies in a WaveNet-like approach. They seem to be doing something along these lines [at MTG](https://mtg.github.io/singing-synthesis-demos/) but I couldn't find any source code (yet). Although, my guess is that if they did release code it would be GPL.

Based on all of this, I can only agree with Kanru's suggestion above that something based on Flite + hts_engine might be the way to go for now, although this would sound nowhere near as good as the current state-of-the-art in terms of quality.

P.S. I'm not planning a Csound Opcode as such, but came across your thread whilst researching a similar project.

**Comment by GloomyGhost-MosquitoCoil** [comment]
@jamiebullock 
Hi there.
I am working in new NN based Singing Synthesizer which is both Parameter synthesis and splicing synthesis, Use the cross over singing model to synthesis the wave and use the nn to optimization the sound quality. and the result seem really effective.

**Comment by gogins** [comment]
Thanks for the information. I'll check ir ak out.

Regards,
Mike

**Comment by gogins** [comment]
I've done some checking. A commercial VST would run in Csound via the vst4cs opcodes on Linux, and might be musically useful, so I'll try Synthesizer V. I'd certainly pay for it if it is useful. GPL code can be used to make a Csound opcode in a separate repository, so I could experiment with MAGE.

**Comment by gogins** [comment]
For Synthesizer-V, libcurl required by it conflicts with many installed packages. Will try just the VST plugin.

**Comment by gogins** [comment]
I will try https://github.com/numediart/mage in a separate repository as with the Aeolus opcodes for Csound.

**Comment by gogins** [comment]
Looking at https://synsig.org/index.php/Software.

**Comment by gogins** [comment]
This uses TensorFlow for speech and is quite impressive even though not musically usable: https://github.com/Kyubyong/tacotron.

**Comment by GloomyGhost-MosquitoSeal** [comment]
I think
Combining the stitching synthesis with the NN vocoder will work well. I am trying this, if possible, I will post the details.

**Comment by gogins** [comment]
This one is even more impressive and would be musically usable if it could sing: https://github.com/sotelo/parrot.

**Comment by gogins** [comment]
DeepMind's [WaveNet](https://deepmind.com/blog/wavenet-generative-model-raw-audio/) is a breakthrough, and is interchangeable with human speech. But, it is not singing. However, it has sparked a great deal of follow on research. It is also, no surprise, very very compute-intensive.

**Comment by gogins** [comment]
This is a kind of hybrid approach that verges on musical usability but source code is not available: https://mtg.github.io/singing-synthesis-demos/.

**Comment by gogins** [comment]
# Restatement of Goal

After reviewing some current research, I restate the project goal.

The objective is to enable Csound to synthesize musically usable, or even human-interchangeable, singing voices based on text input in the Csound score (at best, text with musical, stylistic and expressive markup; at worse, phonemes with musical and expressive markup). It should be possible to generate vocals with no or minimal training data; if training data is necessarily it must be in some widely used or standard format.

**Comment by andiejs** [comment]
"pink trombone" may be relevant/useful
https://dood.al/pinktrombone/
https://github.com/PaulBatchelor/voc
https://github.com/jamesstaub/pink-trombone-osc
https://github.com/zakaton/Pink-Trombone

**Comment by gogins** [comment]
Thanks for the information.

Regards,
Mike

**Comment by gogins** [comment]
Stuff that's open source is still not complete, stuff that actually works is still proprietary. Closing this again.

**Comment by bredfern** [comment]
I've done some things for employers with web tts and ssml  https://www.w3.org/TR/speech-synthesis11/

**Comment by gogins** [comment]
Might check this out: https://n3utrino.work/

**Comment by gogins** [comment]
[This popped up on my screen today. ](https://github.com/r9y9/nnsvs/tree/master/docs)

**Comment by gogins** [comment]
The [Dreamtonics](https://store.dreamtonics.com/) plugin got a decent review in _Sound on Sound_, so I'm re-opening this issue.

The Solaria mezzo voice sounds like it might work.

**Comment by gogins** [comment]
This issue has been open for eight years. There are now a number of singing synthesis VST plugins that are available, some of which have been well reviewed. I will look at these and try using one or more.

https://soundand.design/top-vocal-synthesis-plugins-for-creating-abstract-and-expressive-vocals-in-your-daw-18bbf10e714c

In addition to those mentioned in the link above, there is ACE Studio.

The singing in the demos for Synthesizer V, ACE Studio, and Vocaloid would be usable in my project.

**Comment by gogins** [comment]
The objective here remains setting Emily Dickinson's poem "Safe in their Alabaster Chambers":

> Safe in their Alabaster Chambers —
> Untouched by Morning — and untouched by Noon —
> Lie the meek members of the Resurrection —
> Rafter of Satin — and Roof of Stone!
> 
> Grand go the Years — in the Crescent — above them —
> Worlds scoop their Arcs —
> And Firmaments — row —
> Diadems — drop —
> And Doges — surrender —
> Soundless as Dots,
> On a Disc of Snow.

**Comment by gogins** [comment]
The workflow that I would prefer would not require me personally to sing. I would play in, score, or generate a vocal line, add words to it, and the software would produce audio that is singing the words. How this would happen with Csound is not clear.

Possibly, I could use a score editor and the plugin to type in words for an existing MIDI track, save that as a MIDI file, and the plugin running headless in Csound would still be able to render the singing.

I suppose I should start with the demo version of ACE Studio. No, after looking them over a little more closely, Synthesizer V seems a better starting point, it only costs $89 and the profit to the company comes with addons and more voices.

Drawback: The plugin is designed to run in DAWs using an ARA bridge. I don't think Csound supports the ARA bridge. The [ARA SDK](https://github.com/Celemony/ARA_SDK/blob/main/README.md) has the Apache 2 license. Reaper implements ARA.

However, it looks like you can still create singing tracks with a VST3 Instrument plugin version of the software.

The question arises, can the VSTi version accept real-time MIDI that contains lyrics and inflections? If so the whole shebang can run right in Csound.

Another solution is to create the vocal in Reaper, but load the preset when rendering in Csound; this will work if the preset contains the notes, words, and inflections. Edits in the app are saved in .svp files, I don't know if these are linked to from, or embedded in, presets saved with DAW projects.

**Comment by gogins** [comment]
Possibilities:

1. Add the VST3 GUI to the Csound opcodes.
2. Revive CsoundVST as a VST3 plugin -- but that too would require a GUI!
3. Use Cabbage...? Then I'm composing in a DAW that's running a Csound plugin, the Synthesizer V plugin, and maybe some JavaScript algorithmic composition stuff. Quite a mashup. It would be OK if a Cabbage plugin could edit its .csd file.
4. Then there is singing voice synthesis in WebAssembly. Then the mashup is not so bad. So far, I don't see anything usable.

**Comment by gogins** [comment]
- [x] Purchase Dreamtonics Synthesizer V with free voice.
- [ ] Create a Cabbage instrument that can run any .csd file.
- [ ] Create a workflow in Reaper.
  - [ ] Find out how to enter singing using the Dreamtonics plugin in Reaper.
  - [ ] Find out to edit the Csound orchestra in Reaper.
      - [ ] Cabbage? May be possible -- with difficulty.
      - [ ] Revive vst4cs?
      - [ ] New version of CsoundVST, CsoundVst3, with editor, implemented using JUCE? I have reviewed JUCE 8's license and it is compatible with Csound's license. CsoundVst3 would not be used as software that creates software, so does not conflict with the new part of the JUCE 8 license that forbids that. Cabbage does conflict, but that doesn't appear to affect me.
  - [ ] Find out how to do algorithmic composition in Reaper.
    - [x] See if Reaper's ReaScript engine can import the chord space group in some form. Lua 5.4 is embedded in Reaper.  _However CsoundAC's Lua bindings have been removed._ 
    - [x] The Python bindings are built, are up to date, and should be available in Reaper.
    - [ ] See if the cxx opcodes can be used in Reaper.
    - [ ] See if Csound's `bin` attribute for the `<CsScore>` element can work from Cabbage in Reaper.

**Comment by gogins** [comment]
From ChatGPT:
```
import reaper_python as RPR  # Reaper's Python API

# Function to create a MIDI item and add notes
def create_midi_notes():
    # Get the current track (assumes the first track)
    track = RPR.GetTrack(0, 0)
    if not track:
        print("No track found!")
        return
    
    # Create a new MIDI item on the track
    start_time = 0  # Start time in seconds
    end_time = 4    # End time in seconds
    midi_item = RPR.CreateNewMIDIItemInProj(track, start_time, end_time, False)
    
    # Get the MIDI editor for the item
    midi_take = RPR.GetMediaItemTake(midi_item, 0)
    
    # Add some notes to the MIDI item
    notes = [
        (0, 0.5, 60, 100),  # (start, length, pitch, velocity)
        (0.5, 0.5, 62, 100),
        (1, 0.5, 64, 100),
        (1.5, 0.5, 65, 100),
    ]
    
    for note in notes:
        start, length, pitch, velocity = note
        RPR.MIDI_InsertNote(
            midi_take,    # Take
            False,        # Selected
            False,        # Muted
            RPR.MIDI_GetPPQPosFromProjTime(midi_take, start),              # Start time (PPQ)
            RPR.MIDI_GetPPQPosFromProjTime(midi_take, start + length),     # End time (PPQ)
            0,            # Channel
            pitch,        # Note pitch
            velocity,     # Velocity
            True          # NoSort
        )
    
    # Commit changes
    RPR.MIDI_Sort(midi_take)
    print("MIDI notes added successfully!")

# Run the function
create_midi_notes()
```

**Comment by gogins** [comment]
The above can be packaged as a function to enable an entire `csound::Score` to transfer itself to a MIDI item in Reaper. Instead of notes as a list of tuples, iteration would be over the csound::Notes in the csound::Score.

**Comment by gogins** [comment]
Putting a large .csd file into a Cabbage AUi plugin on a track is possible. The .csd begins playing immediately when the project is loaded. That means, no notes in the .csd file, they have to come from the DAW. Also, inspecting the plugin UI shows no way to get a handle on the text of the .csd file.

There is something fundamentally awkward about this. With this design, it would not at all be easy, may even be impossible, to e.g. save a .csd in a text editor widget and send it to Csound.

**Comment by gogins** [comment]
Plugin _instrument_ formats supported by Reaper on MacOS:

- [x] VSTi, _don't use, no new licenses_!
- [ ] VST3i, _feasible_, check how it works with JUCE.
- [x] AUi, _don't use, only for macOS._
- [x] LV2i, _don't use, complex and mostly for Linux_.
- [ ] CLAPi, _may be better than VST3i_, much better license, and intelligible code also.
- [ ] IPlug2, _supports VST3 and AU_.

**Comment by gogins** [comment]
My options for editing Csound orchestras in Reaper appear to be:

- [ ] Cabbage with workaround for storing .csd in a text editor. Won't support notes stored in the .csd, but that may not matter. Not sure how to stop Csound and recompile the .csd after edits.
- [ ] New CsoundVst3 plugin:
  - [ ] With JUCE 8, which looks more than adequate.
  - [ ] With iPlug2.

**Comment by gogins** [comment]
Steps:

- [ ] Get CsoundAC generating notes in Reaper.
- [ ] Get a Csound orchestra rendering those notes using a Cabbage-generated plugin.
- [ ] Use Synthesizer V to vocalise those notes on a dedicated track or channel.
- [ ] If this is working out, create a bare bones VST3 instrument for Csound, CsoundPlugin.vst3.

**Comment by andiejs** [comment]
I've worked with Synthesizer V through Supercollider and through Reaper. I found I needed to write a .vstpreset file for Synthesizer V in order to load a score of notes / parameters etc. If I were you I'd check that your proposed output format (midi notes in the DAW?) works as input to Synthesizer V before you assume it works the way most other VSTis do. It may have changed since I implemented my stuff a few years ago.

**Comment by gogins** [comment]
Thanks for your comment. I'm aware that the tune and lyrics and aspirations have to be added in the Synthesizer V editor, but the editor can be opened using the FX button in Reaper. 

What was your project?

**Comment by gogins** [comment]
Python3.13 crashes Reaper, Python3.12 works, but CsoundAC is installed for Python3.13. If I run Python3.13 and import CsoundAC, I don't get the latest stuff, it's an earlier build.

But that's easy to fix.

**Comment by andiejs** [comment]
Sounds like you've got a handle on it!
My project is just to wire things together for my creative partners who are composers. Our own script / score language to output files for reaper, supercollider, etc (we trialled csound for a while before settling on supercollider, which I guess is why I'm being alerted on this thread...). We don't have to click things in to the Synthesizer V editor because the file format is basically plaintext json, so we can generate it!
They haven't got deep into composing with synthesizer V yet, but demos sound good. Seems like a pretty powerful system. Hope you have fun with it.

**Comment by gogins** [comment]
Why SuperCollider rather than Csound?

**Comment by gogins** [comment]
I'm closing this because I have bought Synthesizer V, upon which I surely could not improve.

I am moving the Reaper workflow to a new issue.

**Comment by andiejs** [comment]
Hi Michael, sorry I failed to respond and am now replying on a closed thread... I spent a couple days trying to remember why we ended up using supercollider rather than csound and didn't really come up with a reason (and then forgot), but I'm revisiting some of my csound code (which is still hooked up and functional with the rest of my system) and I now have an answer for you.
In both cases, I had to write an interface from our main language to csound and then later supercollider, and the csound one turned out very clunky and unnatural -- mostly because the csound score format is very different from our language, which has functions very much like the supercollider synthdef format, but also partly because I was less experienced with writing language interfaces at the time and didn't do the most beautiful job -- it was very clunky to use. Supercollider had it's own challenges but ultimately that interface is much smoother for us.
I think another factor as to why I was playing with supercollider when I already had csound hooked up (which led to hooking supercollider up with our system) -- I was trying to learn about synthesis, and I found their IDE and the interactiveness of their documentation helped a lot!
There have been a few times when I've needed to look at how both csound and supercollider handle something, and it's always fun to see the different approaches. :-)
